packages/web-wallet/src/contexts/hooks/useWalletConnection.ts

```diff
@@ -0,0 +1,502 @@
+      setLoadingStep('');
+      onError(userMessage);
+    } finally {
+      isCheckingRef.current = false;
+      setIsCheckingConnection(false);
+      setLoadingStep('');
+    }
+  };
+
+  const connectWallet = async () => {
+    setIsConnecting(true);
+    setLoadingStep('Requesting snap...');
+    onError(null);
+
+    try {
+      await requestSnap();
+      setLoadingStep('Checking snap network...');
+
+      let currentSnapNetwork;
+      try {
+        const networkCheckPromise = invokeSnap({
+          method: 'htr_getConnectedNetwork',
+          params: {}
+        });
+        const networkCheckTimeout = new Promise((_, reject) =>
+          setTimeout(() => reject(new Error('Network check timeout')), SNAP_TIMEOUTS.NETWORK_CHECK)
+        );
+        const networkTest = await Promise.race([networkCheckPromise, networkCheckTimeout]);
+
+        let parsedNetworkTest;
+        try {
+          parsedNetworkTest = typeof networkTest === 'string' ? JSON.parse(networkTest) : networkTest;
+        } catch (parseError) {
+          console.error('Failed to parse network test response:', parseError, networkTest);
+          throw new Error(`Snap returned invalid JSON response: ${parseError instanceof Error ? parseError.message : 'Parse error'}`);
+        }
+        currentSnapNetwork = parsedNetworkTest?.response?.network;
+      } catch (testError) {
+        console.error(testError);
+        throw new Error('Snap is not responding. Please make sure it is installed correctly.');
+      }
+
+      const targetNetwork = DEFAULT_NETWORK;
+      if (currentSnapNetwork !== targetNetwork) {
+        setLoadingStep(`Changing snap network to ${DEFAULT_NETWORK}...`);
+
+        try {
+          const changeNetworkPromise = invokeSnap({
+            method: 'htr_changeNetwork',
+            params: {
+              network: currentSnapNetwork,
+              newNetwork: targetNetwork
+            }
+          });
+          const changeNetworkTimeout = new Promise((_, reject) =>
+            setTimeout(() => reject(new Error('Network change timeout')), SNAP_TIMEOUTS.NETWORK_CHANGE)
+          );
+          await Promise.race([changeNetworkPromise, changeNetworkTimeout]);
+        } catch (networkError) {
+          console.error(networkError);
+          throw new Error(`Failed to change snap network to ${DEFAULT_NETWORK}`);
+        }
+      }
+
+      setLoadingStep('Getting xpub from snap...');
+
+      let xpubResponse;
+      try {
+        const xpubPromise = invokeSnap({
+          method: 'htr_getXpub',
+          params: {}
+        });
+        const xpubTimeout = new Promise((_, reject) =>
+          setTimeout(() => reject(new Error('Get xpub timeout')), SNAP_TIMEOUTS.RPC_CALL)
+        );
+        xpubResponse = await Promise.race([xpubPromise, xpubTimeout]);
+      } catch (snapError) {
+        throw new Error(`Failed to get xpub: ${snapError instanceof Error ? snapError.message : String(snapError)}`);
+      }
```

Comment:
We have a lot of races that could be better structured

Reply:
Refactored them in 4fc0b319d1fcc2113885d1157887d37e41ed380f

Much better, thanks